Tasks/FtpUpload/ftpuploadtask.ts

```diff
@@ -0,0 +1,278 @@
+/// <reference path="../../definitions/vsts-task-lib.d.ts" />
+/// <reference path="../../definitions/node.d.ts" />
+/// <reference path="../../definitions/Q.d.ts" />
+
+import fs = require('fs');
+import os = require('os');
+import path = require('path');
+import tl = require('vsts-task-lib/task');
+import url = require('url');
+
+var SortedSet = require('collections/sorted-set');
+var Client = require('ftp');
+
+var win = os.type().match(/^Win/);
+tl.debug('win: ' + win);
+
+var repoRoot: string = tl.getVariable('build.sourcesDirectory');
+function makeAbsolute(normalizedPath: string): string {
+    tl.debug('makeAbsolute:' + normalizedPath);
+
+    var result = normalizedPath;
+    if (!path.isAbsolute(normalizedPath)) {
+        result = path.join(repoRoot, normalizedPath);
+        tl.debug('Relative file path: ' + normalizedPath + ' resolving to: ' + result);
+    }
+    return result;
+}
+
+// server endpoint
+var serverEndpoint = tl.getInput('serverEndpoint', true);
+var serverEndpointUrl : url.Url = url.parse(tl.getEndpointUrl(serverEndpoint, false));
+
+var serverEndpointAuth = tl.getEndpointAuthorization(serverEndpoint, false);
+var username = serverEndpointAuth['parameters']['username'];
+var password = serverEndpointAuth['parameters']['password'];
+
+// the root location which will be uploaded from
+var rootFolder: string = makeAbsolute(path.normalize(tl.getPathInput('rootFolder', true).trim()));
+if (!tl.exist(rootFolder)) {
+    failTask('The specified root folder does not exist: ' + rootFolder );
+}
+
+var clean: boolean = tl.getBoolInput('clean', true);
+var overwrite: boolean = tl.getBoolInput('overwrite', true);
+var preservePaths: boolean = tl.getBoolInput('preservePaths', true);
+
+function findFiles(): string[] {
```

Comment:
It seems wrong to write all this code in a task.  What's missing in the libs match, filter and glob methods?